File: structures/queue/queue.go
package main

import "fmt"

// 用两个栈模拟队列

// Stack 定义栈结构
type Stack struct {
	elements []int
}

// Push 向栈中推入一个元素
func (s *Stack) Push(element int) {
	s.elements = append(s.elements, element)
}

// Pop 从栈中弹出一个元素
func (s *Stack) Pop() int {
	if len(s.elements) == 0 {
		return -1
	}
	element := s.elements[len(s.elements)-1]
	s.elements = s.elements[:len(s.elements)-1]
	return element
}

// IsEmpty 判断栈是否为空
func (s *Stack) IsEmpty() bool {
	return len(s.elements) == 0
}

// Queue 用两个栈模拟队列
type Queue struct {
	inStack  Stack // 用来入队
	outStack Stack // 用来出队
}

// Enqueue 入队
func (q *Queue) Enqueue(element int) {
	q.inStack.Push(element)
}

// Dequeue 出队 如果outStack不为空，则直接从outStack中弹出一个元素
// 如果outStack为空，则从inStack中弹出所有元素，然后依次入outStack
func (q *Queue) Dequeue() int {
	if q.outStack.IsEmpty() {
		for !q.inStack.IsEmpty() {
			q.outStack.Push(q.inStack.Pop())
		}
	}
	return q.outStack.Pop()
}

func main() {
	queue := &Queue{}
	queue.Enqueue(1)
	queue.Enqueue(2)
	queue.Enqueue(3)
	a := queue.Dequeue()
	fmt.Println(a)
}

File: structures/queue/queue.rs
use std::vec;

struct Stack{
    elements:Vec<i32>
}
impl Stack{
    fn new()->Stack{
        Stack{
            elements:vec![]
        }
    }
    // 向栈中推入一个元素
    fn push(&mut self,element:i32){
        self.elements.push(element);
    }
    // 从栈中弹出一个元素
    fn pop(&mut self)->Option<i32>{
        self.elements.pop()
    }
    // 栈是否为空
    fn is_empty(&self)->bool{
        self.elements.is_empty()
    }
}

// 用两个栈模拟队列
struct Queue{
    in_stack:Stack, // 用来模拟入队
    out_stack:Stack // 用来模拟出队
}
impl Queue{
    fn new()->Queue{
        Queue{
            in_stack:Stack::new(),
            out_stack:Stack::new()
        }
    }
    // 入队
    fn enqueue(&mut self,val:i32){
        self.in_stack.push(val)
    }
    // 出队
    fn dequeue(&mut self)->Option<i32>{
        if self.out_stack.is_empty(){
            while !self.in_stack.is_empty(){
                let val = self.in_stack.pop().unwrap(); // 因为rust的pop()返回Option<T>类型，所以需要unwrap()
                self.out_stack.push(val);
            }
        }
        self.out_stack.pop()
    }

}

fn main(){
    let mut queue = Queue::new();
    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);
    let a=queue.dequeue();
    println!("{:?}",a)

}